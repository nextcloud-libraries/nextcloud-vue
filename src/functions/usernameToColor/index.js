/**
 * SPDX-FileCopyrightText: 2018 Nextcloud GmbH and Nextcloud contributors
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import { GenColors } from '../../utils/GenColors.js'

import md5 from 'md5'

/**
 * Generate a color from a username
 * Originally taken from https://github.com/nextcloud/server/blob/master/core/js/placeholder.js
 *
 * @param {string} username Display name or user id to generate from
 * @return {{ r: number, g: number, b: number }} the RGB color
 */
export function usernameToColor(username) {
	// Normalize hash
	let hash = username.toLowerCase()

	// Already a md5 hash?
	if (hash.match(/^([0-9a-f]{4}-?){8}$/) === null) {
		hash = md5(hash)
	}

	hash = hash.replace(/[^0-9a-f]/g, '')

	const steps = 6
	const finalPalette = GenColors(steps)

	/**
	 * Convert a string to an integer evenly
	 *
	 * @param {string} hash The hash to convert
	 * @param {number} maximum Largest number allowed
	 */
	function hashToInt(hash, maximum) {
		let finalInt = 0
		const result = []

		// Splitting evenly the string
		for (let i = 0; i < hash.length; i++) {
			// chars in md5 goes up to f, hex:16
			result.push(parseInt(hash.charAt(i), 16) % 16)
		}

		// Adds up all results
		for (const j in result) {
			finalInt += result[j]
		}

		// chars in md5 goes up to f, hex:16
		// make sure we're always using int in our operation
		return parseInt(parseInt(finalInt, 10) % maximum, 10)
	}
	return finalPalette[hashToInt(hash, steps * 3)]
}

/**
 * @deprecated use the named export of `usernameToColor` instead
 */
export default usernameToColor
