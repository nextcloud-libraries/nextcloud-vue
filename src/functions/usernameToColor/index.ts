/**
 * SPDX-FileCopyrightText: 2018 Nextcloud GmbH and Nextcloud contributors
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import type { Color } from '../../utils/colors.ts'

import { Md5 } from 'ts-md5'
import { generatePalette } from '../../utils/colors.ts'

/**
 * Create a simple hash from a string
 *
 * @param str - The string to hash
 */
function hashCode(str: string): number {
	let hash = str

	// Hash a given string, if it is not md5 hash already
	if (str.match(/^([0-9a-f]{4}-?){8}$/) === null) {
		hash = Md5.hashStr(str)
	}

	hash = hash.replace(/[^0-9a-f]/g, '')

	let finalInt = 0

	for (let i = 0; i < hash.length; i++) {
		// chars in md5 are [0-9a-f] (base-16)
		finalInt += parseInt(hash.charAt(i), 16)
	}

	return finalInt
}

/**
 * Generate a color from a username
 *
 * @param username - Display name or user id to generate from
 * @return The RGB color
 */
export function usernameToColor(username: string): Color {
	const steps = 6
	const finalPalette = generatePalette(steps)
	const hash = hashCode(username.toLocaleLowerCase())

	return finalPalette[hash % finalPalette.length]!
}
