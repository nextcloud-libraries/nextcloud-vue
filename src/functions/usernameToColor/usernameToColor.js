/**
 * @copyright Copyright (c) 2018 John Molakvoæ <skjnldsv@protonmail.com>
 *
 * @author John Molakvoæ <skjnldsv@protonmail.com>
 *
 * @license AGPL-3.0-or-later
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 *
 */
import GenColors from '../../utils/GenColors.js'

import md5 from 'md5'

/**
 * Originally taken from https://github.com/nextcloud/server/blob/master/core/js/placeholder.js
 *
 * @param {string} username Display name or user id to generate from
 * @return {object} the rgb colors as {r:255, g:255, b:255}
 */
const usernameToColor = function(username) {
	// Normalize hash
	let hash = username.toLowerCase()

	// Already a md5 hash?
	if (hash.match(/^([0-9a-f]{4}-?){8}$/) === null) {
		hash = md5(hash)
	}

	hash = hash.replace(/[^0-9a-f]/g, '')

	const steps = 6
	const finalPalette = GenColors(steps)

	/**
	 * Convert a string to an integer evenly
	 *
	 * @param {string} hash The hash to convert
	 * @param {number} maximum Largest number allowed
	 */
	function hashToInt(hash, maximum) {
		let finalInt = 0
		const result = []

		// Splitting evenly the string
		for (let i = 0; i < hash.length; i++) {
			// chars in md5 goes up to f, hex:16
			result.push(parseInt(hash.charAt(i), 16) % 16)
		}

		// Adds up all results
		for (const j in result) {
			finalInt += result[j]
		}

		// chars in md5 goes up to f, hex:16
		// make sure we're always using int in our operation
		return parseInt(parseInt(finalInt, 10) % maximum, 10)
	}
	return finalPalette[hashToInt(hash, steps * 3)]
}

export default usernameToColor
