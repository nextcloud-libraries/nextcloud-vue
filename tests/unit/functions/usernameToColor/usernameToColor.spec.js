import { describe, expect, it } from 'vitest'

import usernameToColor from '../../../../src/functions/usernameToColor/index'

const rgbToHex = (color) => {
	return '#' + ((1 << 24) + (color.r << 16) + (color.g << 8) + color.b).toString(16).slice(1)
}

describe('usernameToColor', () => {
	'use strict'

	const usernames = [
		'', ',', '.', 'admin', 'foo', 'bar', 'asd', 'wasd', 'a user', 'another user', 'admin@cloud.example.com',
		'Lorelai Taylor',
		'Lillian Wall',
		'Haseeb Stephens',
		'Rae Hope',
		'Santiago Singleton',
		'Daphne Roy',
		'Arham Johnson',
		'Madina Knight',
		'Sid Combs',
		'Zaki Cortes',
		'Ellena Wright Frederic Conway',
		'Alishia Ann Lowry',
		'Idris Mac',
		'Kristi Fisher',
		'Gianluca Hills',
		'Vivienne Jacobs',
		'Brayden Truong',
		'Akeel Robertson',
		'Tom MÃ¶rtel',
		'ðŸ™ˆ',
		'Ù…Ø±Ø­Ø¨Ø§ Ø¨Ø§Ù„Ø¹Ø§Ù„Ù…',
		'123e4567-e89b-12d3-a456-426614174000',
		'Meeting', 'TestCircle', 'Private Circle',
	]

	usernames.forEach((key) => {
		it(key + ' has the proper color', () => {
			expect(rgbToHex(usernameToColor(key))).toMatchSnapshot()
		})
	})

})
