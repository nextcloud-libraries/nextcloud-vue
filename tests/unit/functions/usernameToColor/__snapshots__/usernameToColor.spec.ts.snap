// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`usernameToColor >  has the proper color 1`] = `"#0082c9"`;

exports[`usernameToColor > , has the proper color 1`] = `"#1e78c1"`;

exports[`usernameToColor > . has the proper color 1`] = `"#c98879"`;

exports[`usernameToColor > 123e4567-e89b-12d3-a456-426614174000 has the proper color 1`] = `"#bc5c91"`;

exports[`usernameToColor > Akeel Robertson has the proper color 1`] = `"#9750a4"`;

exports[`usernameToColor > Alishia Ann Lowry has the proper color 1`] = `"#d09e6d"`;

exports[`usernameToColor > Arham Johnson has the proper color 1`] = `"#0082c9"`;

exports[`usernameToColor > Brayden Truong has the proper color 1`] = `"#d09e6d"`;

exports[`usernameToColor > Daphne Roy has the proper color 1`] = `"#9750a4"`;

exports[`usernameToColor > Ellena Wright Frederic Conway has the proper color 1`] = `"#c37285"`;

exports[`usernameToColor > Gianluca Hills has the proper color 1`] = `"#d6b461"`;

exports[`usernameToColor > Haseeb Stephens has the proper color 1`] = `"#d6b461"`;

exports[`usernameToColor > Idris Mac has the proper color 1`] = `"#9750a4"`;

exports[`usernameToColor > Kristi Fisher has the proper color 1`] = `"#0082c9"`;

exports[`usernameToColor > Lillian Wall has the proper color 1`] = `"#bc5c91"`;

exports[`usernameToColor > Lorelai Taylor has the proper color 1`] = `"#ddcb55"`;

exports[`usernameToColor > Madina Knight has the proper color 1`] = `"#9750a4"`;

exports[`usernameToColor > Meeting has the proper color 1`] = `"#c98879"`;

exports[`usernameToColor > Private Circle has the proper color 1`] = `"#c37285"`;

exports[`usernameToColor > Rae Hope has the proper color 1`] = `"#795aab"`;

exports[`usernameToColor > Santiago Singleton has the proper color 1`] = `"#bc5c91"`;

exports[`usernameToColor > Sid Combs has the proper color 1`] = `"#d09e6d"`;

exports[`usernameToColor > TestCircle has the proper color 1`] = `"#499aa2"`;

exports[`usernameToColor > Tom MÃ¶rtel has the proper color 1`] = `"#248eb5"`;

exports[`usernameToColor > Vivienne Jacobs has the proper color 1`] = `"#1e78c1"`;

exports[`usernameToColor > Zaki Cortes has the proper color 1`] = `"#6ea68f"`;

exports[`usernameToColor > a user has the proper color 1`] = `"#5b64b3"`;

exports[`usernameToColor > admin has the proper color 1`] = `"#d09e6d"`;

exports[`usernameToColor > admin@cloud.example.com has the proper color 1`] = `"#9750a4"`;

exports[`usernameToColor > another user has the proper color 1`] = `"#ddcb55"`;

exports[`usernameToColor > asd has the proper color 1`] = `"#248eb5"`;

exports[`usernameToColor > bar has the proper color 1`] = `"#0082c9"`;

exports[`usernameToColor > foo has the proper color 1`] = `"#d09e6d"`;

exports[`usernameToColor > wasd has the proper color 1`] = `"#b6469d"`;

exports[`usernameToColor > Ù…Ø±Ø­Ø¨Ø§ Ø¨Ø§Ù„Ø¹Ø§Ù„Ù… has the proper color 1`] = `"#c98879"`;

exports[`usernameToColor > ğŸ™ˆ has the proper color 1`] = `"#b6469d"`;
